src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQueryPullDataset.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.gcp.bigquery.sqlengine;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.Table;
+import com.google.cloud.bigquery.TableId;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.dataset.lib.KeyValue;
+import io.cdap.cdap.etl.api.Emitter;
+import io.cdap.cdap.etl.api.SerializableTransform;
+import io.cdap.cdap.etl.api.Transform;
+import io.cdap.cdap.etl.api.engine.sql.dataset.SQLPullDataset;
+import io.cdap.cdap.etl.api.engine.sql.request.SQLPullRequest;
+import io.cdap.plugin.gcp.bigquery.source.BigQueryInputFormatProvider;
+import io.cdap.plugin.gcp.bigquery.source.BigQuerySourceUtils;
+import io.cdap.plugin.gcp.bigquery.util.BigQueryConstants;
+import org.apache.avro.generic.GenericData;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.LongWritable;
+
+import java.io.IOException;
+import javax.annotation.Nullable;
+
+/**
+ * SQL Pull Dataset implementation for BigQuery backed datasets.
+ */
+public class BigQueryPullDataset extends BigQueryInputFormatProvider
+  implements SQLPullDataset<StructuredRecord, LongWritable, GenericData.Record>, BigQuerySQLDataset {
+
+  // Note that these properties are not final in order to support Spark's serialization.
+  private transient BigQuery bigQuery;
+  private String datasetName;
+  private Schema schema;
+  private String project;
+  private String bqDataset;
+  private String bqTable;
+  private String gcsPath;
+  private Long numRows;
+  private final SQLEngineAvroToStructuredTransformer transformer = new SQLEngineAvroToStructuredTransformer();
```

Comment:
since this is used within the anonymous function returned by fromKeyValue(), Spark needs to serialize this entire class. It would be better to pull that Transform into its own class and move this variable to that class. That way BigQueryPullDataset doesn't need to be serializable, and this becomes more straightforward to understand.

Reply:
Yep, that worked.